kubespawner/proxy.py

```diff
@@ -0,0 +1,244 @@
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+        'hub.jupyter.org/proxy-route': 'true'
+    }
+
+    list_method_name = 'list_namespaced_ingress'
+    api_group_name = 'ExtensionsV1beta1Api'
+
+    @property
+    def ingresses(self):
+        return self.resources
+
+class ServiceReflector(NamespacedResourceReflector):
+    kind = 'services'
+    labels = {
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+        'hub.jupyter.org/proxy-route': 'true'
+    }
+
+    list_method_name = 'list_namespaced_service'
+
+    @property
+    def services(self):
+        return self.resources
+
+class EndpointsReflector(NamespacedResourceReflector):
+    kind = 'endpoints'
+    labels = {
+        'heritage': 'jupyterhub',
+        'component': 'singleuser-server',
+        'hub.jupyter.org/proxy-route': 'true'
+    }
+
+    list_method_name = 'list_namespaced_endpoints'
+
+    @property
+    def endpoints(self):
+        return self.resources
+
+class KubeIngressProxy(Proxy):
+    namespace = Unicode(
+        config=True,
+        help="""
+        Kubernetes namespace to spawn ingresses for single-user servers in.
+
+        If running inside a kubernetes cluster with service accounts enabled,
+        defaults to the current namespace. If not, defaults to 'default'
+        """
+    )
+
+    def _namespace_default(self):
+        """
+        Set namespace default to current namespace if running in a k8s cluster
+
+        If not in a k8s cluster with service accounts enabled, default to
+        'default'
+        """
+        ns_path = '/var/run/secrets/kubernetes.io/serviceaccount/namespace'
+        if os.path.exists(ns_path):
+            with open(ns_path) as f:
+                return f.read().strip()
+        return 'default'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # We want 3x concurrent spawn limit as our threadpool. This means that if
+        # concurrent_spawn_limit servers start instantly, we can add all 3 objects
+        # required to add proxy routing for them instantly as well.
+        self.executor = ThreadPoolExecutor(max_workers=self.app.concurrent_spawn_limit * 3)
+
+        self.ingress_reflector = IngressReflector(parent=self, namespace=self.namespace)
+        self.service_reflector = ServiceReflector(parent=self, namespace=self.namespace)
+        self.endpoint_reflector = EndpointsReflector(parent=self, namespace=self.namespace)
+
+        self.core_api = client.CoreV1Api()
+        self.extension_api = client.ExtensionsV1beta1Api()
+
+    @run_on_executor
+    def asynchronize(self, method, *args, **kwargs):
+        return method(*args, **kwargs)
+
+    def safe_name_for_routespec(self, routespec):
+        safe_chars = set(string.ascii_lowercase + string.digits)
+        safe_name = generate_hashed_slug(
+            'jupyter-' + escapism.escape(routespec, safe=safe_chars, escape_char='-') + '-route'
```

Comment:
Should we make this configurable? We could also pre-replace all '/' with -, since otherwise the objects look kinda unweildy, like 'jupyter--2fuser-2ftester-2f-route'.

Reply:
Making escape_char configurable is probably a good idea. I noticed in binderhub that since we are creating users with `-` in their names, they are getting pods called `jupyter-2dfoo-2d`, etc. This escaping `-` is technically necessary to rigorously avoid collisions, but not necessary in the binderhub case in particular, where we can be sure that there won't be collisions.

Reply:
+1, I'll make that happen here and on the spawner too.